sdk/lib/conutils/pager.c

```diff
@@ -44,74 +226,65 @@ ConWritePaging(
-    DWORD from = 0, i = 0;
-
     /* Parameters validation */
     if (!Pager)
         return FALSE;
 
-    /* Reset LineCount and return if no string has been given */
-    if (StartPaging == TRUE)
-        Pager->LineCount = 0;
-    if (szStr == NULL)
-        return TRUE;
+    if (StartPaging)
+    {
+        /* Reset the output line count, the column index and the line number */
+        Pager->iLine = 0;
+        Pager->iColumn = 0;
+        Pager->lineno = 1;
+        Pager->nSpacePending = 0;
+    }
 
     /* Get the size of the visual screen that can be printed to */
     if (!ConGetScreenInfo(Pager->Screen, &csbi))
     {
         /* We assume it's a file handle */
-        CON_STREAM_WRITE(Pager->Screen->Stream, szStr, len);
+        ConCallPagerLine(Pager, szStr, len);
         return TRUE;
     }
 
-    /*
-     * Get the number of lines currently displayed on screen, minus 1
-     * to account for the "press any key..." prompt from PagePrompt().
-     */
-    ScreenLines = (csbi.srWindow.Bottom - csbi.srWindow.Top);
-    CharSL = csbi.dwCursorPosition.X;
+    /* Fill the pager info */
+    Pager->ScreenColumns = csbi.srWindow.Right - csbi.srWindow.Left + 1;
+    Pager->ScreenRows = csbi.srWindow.Bottom - csbi.srWindow.Top + 1;
+    Pager->ich = 0;
+    Pager->cch = len;
+    Pager->TextBuff = szStr;
+    if (StartPaging)
+    {
+        /* Reset to display one page by default */
+        Pager->ScrollRows = Pager->ScreenRows - 1;
+    }
+
+    if (len == 0 || szStr == NULL)
+        return TRUE;
 
     /* Make sure the user doesn't have the screen too small */
-    if (ScreenLines < 4)
+    if (Pager->ScreenRows < 4)
     {
-        CON_STREAM_WRITE(Pager->Screen->Stream, szStr, len);
+        ConCallPagerLine(Pager, szStr, len);
         return TRUE;
     }
 
-    while (i < len)
+    while (!ConPagerDefaultAction(Pager))
     {
-        /* Search until the end of a line is reached */
-        if (szStr[i++] != TEXT('\n') && ++CharSL < csbi.dwSize.X)
-            continue;
-
-        Pager->LineCount++;
-        CharSL = 0;
-
-        if (Pager->LineCount >= ScreenLines)
+        /* Recalculate the screen extent in case the user redimensions the window */
+        if (ConGetScreenInfo(Pager->Screen, &csbi))
         {
-            CON_STREAM_WRITE(Pager->Screen->Stream, &szStr[from], i-from);
-            from = i;
-
-            /* Prompt the user; give him some values for statistics */
-            // FIXME TODO: The prompt proc can also take ScreenLines ??
-            if (!PagePrompt(Pager, from, len))
-                return FALSE;
+            Pager->ScreenColumns = csbi.srWindow.Right - csbi.srWindow.Left + 1;
+            Pager->ScreenRows = csbi.srWindow.Bottom - csbi.srWindow.Top + 1;
+        }
 
-            // TODO: Recalculate 'ScreenLines' in case the user redimensions
-            // the window during the prompt.
+        /* Reset to display one page by default */
+        Pager->ScrollRows = Pager->ScreenRows - 1;
```

Comment:
I find this: resetting the ScrollRows value to its default, to be advantageous, so that the PagePrompt callback doesn't need to care about the "correct" value to set there, when it wants to tell the pager to display a full screen page. The PagePrompt callback resets the ScrollRows value to a different value only when it wants, for the next paging action, to change the number of lines.